File: src/main/java/org/runestar/cache/Compressor.java
package org.runestar.cache;

import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream;
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorOutputStream;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.SequenceInputStream;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

public enum Compressor {

    NONE((byte) 0, 0),

    BZIP2((byte) 1, Integer.BYTES) {

        private final int BLOCK_SIZE = 1;

        private final byte[] HEADER = ("BZh" + BLOCK_SIZE).getBytes(StandardCharsets.US_ASCII);

        @Override
        protected ByteBuffer compress0(ByteBuffer buf) {
            var decompressedSize = buf.getInt();
            var output = new ByteArrayOutputStream();
            try (var in = new ByteBufferInputStream(buf);
                 var out = new BZip2CompressorOutputStream(output, BLOCK_SIZE)) {
                in.transferTo(out);
            } catch (IOException e) {
                throw new IllegalArgumentException(e);
            }
            var outputArray = output.toByteArray();
            if (!Arrays.equals(outputArray, 0, HEADER.length, HEADER, 0, HEADER.length)) {
                throw new IllegalStateException();
            }
            return ByteBuffer.wrap(outputArray).putInt(0, decompressedSize);
        }

        @Override
        protected ByteBuffer decompress0(ByteBuffer buf) {
            var output = new byte[buf.getInt()];
            try (var seq = new SequenceInputStream(new ByteArrayInputStream(HEADER), new ByteBufferInputStream(buf));
                 var in = new BZip2CompressorInputStream(seq)) {
                IO.readNBytes(in, output);
            } catch (IOException e) {
                throw new IllegalArgumentException(e);
            }
            return ByteBuffer.wrap(output);
        }
    },

    GZIP((byte) 2, Integer.BYTES) {

        @Override
        protected ByteBuffer compress0(ByteBuffer buf) {
            var output = new ByteArrayOutputStream();
            try {
                IO.writeInt(output, buf.remaining());
            } catch (IOException e) {
                throw new IllegalStateException();
            }
            try (var in = new ByteBufferInputStream(buf);
                 var out = new GZIPOutputStream(output)) {
                in.transferTo(out);
            } catch (IOException e) {
                throw new IllegalArgumentException(e);
            }
            return ByteBuffer.wrap(output.toByteArray());
        }

        @Override
        protected ByteBuffer decompress0(ByteBuffer buf) {
            var output = new byte[buf.getInt()];
            try (var in = new GZIPInputStream(new ByteBufferInputStream(buf))) {
                IO.readNBytes(in, output);
            } catch (IOException e) {
                throw new IllegalArgumentException(e);
            }
            return ByteBuffer.wrap(output);
        }
    };

    public final byte id;

    public final int headerLength;

    Compressor(byte id, int headerLength) {
        this.id = id;
        this.headerLength = headerLength;
    }

    protected ByteBuffer compress0(ByteBuffer buf) {
        return buf;
    }

    protected ByteBuffer decompress0(ByteBuffer buf) {
        return buf;
    }

    public static Compressor of(byte id) {
        switch (id) {
            case 0: return NONE;
            case 1: return BZIP2;
            case 2: return GZIP;
        }
        throw new IllegalArgumentException(String.valueOf(id));
    }

    public static ByteBuffer decompress(ByteBuffer buf) {
        return decompress(buf, null);
    }

    public static ByteBuffer decompress(ByteBuffer buf, int[] key) {
        var compressor = Compressor.of(buf.get());
        var compressedLength = buf.getInt() + compressor.headerLength;
        var compressedLimit = buf.position() + compressedLength;
        var totalLimit = buf.limit();
        buf.limit(compressedLimit);
        if (key != null) {
            XteaCipher.decrypt(buf, key);
        }
        var decompressed = compressor.decompress0(buf);
        buf.limit(totalLimit);
        return decompressed;
    }
}

File: src/main/java/com/runesuite/cache/Compressor.kt
package com.runesuite.cache

import com.runesuite.cache.extensions.*
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.netty.util.AsciiString
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorOutputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream

enum class Compressor(val id: Int, val headerLength: Int) {

    NONE(0, 0) {
        override fun compress(buffer: ByteBuf): ByteBuf {
            return buffer
        }

        override fun decompress(buffer: ByteBuf): ByteBuf {
            return buffer
        }
    },

    BZIP2(1, Integer.BYTES) {
        private val BLOCK_SIZE = 1

        private val HEADER: AsciiString = "BZh$BLOCK_SIZE".toAscii()

        override fun compress(buffer: ByteBuf): ByteBuf {
            buffer.inputStream().use { input ->
                val bufferOutput = Unpooled.buffer().outputStream()
                bufferOutput.writeInt(buffer.readableBytes())
                BZip2CompressorOutputStream(bufferOutput, BLOCK_SIZE).use { output ->
                    input.copyTo(output)
                    val buf = bufferOutput.buffer()
                    val header = buf.readArray(HEADER.length).asAscii()
                    check(header.contentEquals(HEADER)) { "Invalid header: $header" }
                    return buf
                }
            }
        }

        override fun decompress(buffer: ByteBuf): ByteBuf {
            val expectedDecompressedSize = buffer.readInt()
            BZip2CompressorInputStream(HEADER.array().inputStream() + buffer.inputStream()).use { input ->
                Unpooled.buffer(expectedDecompressedSize).outputStream().use { output ->
                    input.copyTo(output)
                    val decompressedSize = output.buffer().readableBytes()
                    check(decompressedSize == expectedDecompressedSize)
                    return output.buffer()
                }
            }
        }
    },

    GZIP(2, Integer.BYTES) {
        override fun compress(buffer: ByteBuf): ByteBuf {
            buffer.inputStream().use { input ->
                val bufferOutput = Unpooled.buffer().outputStream()
                bufferOutput.writeInt(buffer.readableBytes())
                GzipCompressorOutputStream(bufferOutput).use { output ->
                    input.copyTo(output)
                    return bufferOutput.buffer()
                }
            }
        }

        override fun decompress(buffer: ByteBuf): ByteBuf {
            val expectedDecompressedSize = buffer.readInt()
            GzipCompressorInputStream(buffer.inputStream()).use { input ->
                Unpooled.buffer(expectedDecompressedSize).outputStream().use { output ->
                    input.copyTo(output)
                    val decompressedSize = output.buffer().readableBytes()
                    check(decompressedSize == expectedDecompressedSize)
                    return output.buffer()
                }
            }
        }
    };

    abstract fun decompress(buffer: ByteBuf): ByteBuf

    abstract fun compress(buffer: ByteBuf): ByteBuf

    companion object {
        val LOOKUP = values().associateBy { it.id }
    }
}